asQ/post.py

```diff
@@ -0,0 +1,148 @@
+
+import firedrake as fd
+
+from pyop2.mpi import MPI
+
+
+def write_timesteps(pdg,
+                    file_name='paradiag_output',
+                    function_names=[],
+                    only_last_step=False):
+    """Writes each timestep of a paradiag object to seperate vtk files.
+
+    :arg pdg: the paradiag object
+    :arg file_name: optional name for the files. The full filename will be file_name.{timestep}
+    :arg function_names: a list of names for each function in the (mixed) function space at each timestep
+    :arg only_last_step: only writes the last timestep and does not use .{timestep} suffix on filename
+    """
+
+    # TODO: This implementation assumes that a MixedFunctionSpace is used at each timestep
+    #       Once there is an example using a plain FunctionSpace this will need updating
+
+    # if given, check we have the right number of function_names
+    if (len(function_names) != 0) and (len(function_names) != pdg.ncpts):
+        raise ValueError("function_names must be same length as pdg.ncpts,"
+                         + f" {len(function_names)} provided, {pdg.ncpts} needed.")
+
+    # functions for writing to file
+    functions = []
+    for cpt in range(pdg.ncpts):
+        V = pdg.W.split()[cpt]
+        if len(function_names) != 0:
+            functions.append(fd.Function(V, name=function_names[cpt]))
+        else:
+            functions.append(fd.Function(V))
+
+    # functions from entire local time-slice
+    walls = pdg.w_all.split()
+
+    # first timestep of this local time-slice
+    timestep0 = sum(pdg.M[:pdg.rT])
+
+    if only_last_step:
+        if pdg.rT == (len(pdg.M) - 1):
+            i = pdg.M[pdg.rT]-1
+            timestep = timestep0+i
+
+            # index of first split function in this timestep
+            index0 = pdg.ncpts*i
+
+            for cpt in range(pdg.ncpts):
+                functions[cpt].assign(walls[index0+cpt])
+
+            fd.File(file_name+".pvd",
+                    comm=pdg.ensemble.comm).write(*functions)
+
+            return
+
+    else:  # write every timestep
+
+        for i in range(pdg.M[pdg.rT]):
+            timestep = timestep0+i
+
+            # index of first split function in this timestep
+            index0 = pdg.ncpts*i
+
+            for cpt in range(pdg.ncpts):
+                functions[cpt].assign(walls[index0+cpt])
+
+            fd.File(file_name+"."+str(timestep)+".pvd",
+                    comm=pdg.ensemble.comm).write(*functions)
+
+
+def write_timeseries(pdg,
+                     file_name='paradiag_output',
+                     function_names=[]):
```

Comment:
please can we have an option that is a halfway house between writing every timestep and only the last one?

We could specify a frequency, and still request all the data but only write some of it. This is not ideal but I want to replace this with proper parallel checkpointing in the near future.

Reply:
Speficying a frequency should be straightforward. I don't think we'd even need to request all the data, just put a frequency on the ranged for loop

Reply:
See commit [4a4d3b6](https://github.com/colinjcotter/asQ/pull/33/commits/4a4d3b668252c5d283fb2f6c695d005e60b5b0ca)